packages/go_router/lib/src/delegate.dart

```diff
@@ -146,15 +160,40 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
     return true;
   }
 
-  /// Replaces the top-most page of the page stack with the given one.
+  /// Replaces the top-most page of the page stack with the given one. The page
+  /// key of the new page will always be different from the old one.
   ///
   /// See also:
   /// * [push] which pushes the given location onto the page stack.
+  /// * [replace] which replaces the top-most page of the page stack but keeps
+  ///   the page key when possible.
   void pushReplacement(RouteMatchList matches) {
+    assert(matches.last.route is! ShellRoute);
     _matchList.remove(_matchList.last);
     push(matches); // [push] will notify the listeners.
   }
 
+  /// Replaces the top-most page of the page stack with the given one and reuse
+  /// the page key when the path didn't change.
+  ///
+  /// See also:
+  /// * [push] which pushes the given location onto the page stack.
+  /// * [pushReplacement] which replaces the top-most page of the page stack but
+  ///   always use a new page key.
+  void replace(RouteMatchList matches) {
```

Comment:
This method starts to deviate from Navigator.replace, which can replace any route instead of just the top one. One thing I think that may be useful would be replace GoRoute with an other GoRoute. maybe we can use GoRoute.name as here? WDYT?

also cc @johnpryan  for API feedback

Reply:
I'm not trying to re-implement `Navigator.replace` API here. The main goal of this PR is to fix https://github.com/flutter/flutter/issues/115902. The solution I came up with was this `replace` method (which we could name differently, ex: `replaceLast`). I don't know whether re-implementing `Navigator.replace` would fix the issue (maybe it will not re-use the same key, like `pushReplacement`) or not.

Having this in mind, what do you think we should do in this PR?

Reply:
My worry is that someone later on may file another feature request to replace the middle of match list, and then we would need to come up with a new api since the function signature here cannot be reused.

I also been thinking about opening up the RouteMatchList and all the package level private class. That may open a better way to manage the RouteMatchList.

So I think there are two ways to go from here. Either we expand this to be able to replace any GoRoute(or a sub list of RouteBases) with one or more RouteBases. Or we open up the API now and let customers to do their things.

Reply:
> So I think there are two ways to go from here. Either we expand this to be able to replace any GoRoute(or a sub list of RouteBases) with one or more RouteBases. Or we open up the API now and let customers to do their things.

Is there a way you prefer over the other? I guess I can try to implement the `replace` method here

Reply:
Hello @chunhtai I tried to look a bit at how it could be possible to make `replace` a bit like

```dart
void Navigator.replace({ required Route<dynamic> oldRoute, required Route<T> newRoute})
```

So I guess, from what you said:

> One thing I think that may be useful would be replace GoRoute with an other GoRoute. maybe we can use GoRoute.name as here

`replace` could have one of the signatures (if `routeNameToReplace` is `null`, it would replace the top most route).

```dart
void replace(RouteMatchList matchese, {String? routeNameToReplace});
void replace(RouteBase newRoute, {String? routeNameToReplace});
```

There are several issues with this:
- If a route is present multiple times in the history, which one is supposed to be replaced?
- This solution doesn't work with `go_router_builder` as all the generated routes are nameless.

---

About your concerns

> My worry is that someone later on may file another feature request to replace the middle of match list, and then we would need to come up with a new api since the function signature here cannot be reused.

The signature of the method `replace` of the delegate could be changed easily as it is not opened yet, so there should not be any issues there.

The signature of the method `replace` of the router should not be breaking. It could change from

```dart
void replace(String location, {Object? extra});
```

to 

```dart
void replace(String location, {WhateverType? whateverParameter, Object? extra});
```

---

What do you think?

Reply:
I am still slightly concern about using a string location to do the replace, but i can't find a better alternative. yeah, I am ok with this approach

Reply:
please re-request me for review when this pr is ready.

Reply:
> yeah, I am ok with this approach

What approach are you okay with? The one currently implemented in this PR ? (If yes, it would be ready to review)

If you are talking about

```dart
void replace(RouteMatchList matchese, {String? routeNameToReplace});
void replace(RouteBase newRoute, {String? routeNameToReplace});
```

What about 

> There are several issues with this:
>
> - If a route is present multiple times in the history, which one is supposed to be replaced?
> - This solution doesn't work with go_router_builder as all the generated routes are nameless.

?

Reply:
the current implementation